scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -26,9 +26,50 @@ import com.spotify.scio.util.random.{BernoulliValueSampler, PoissonValueSampler}
 import com.twitter.algebird._
 import org.apache.beam.sdk.transforms.{Aggregator => _, _}
 import org.apache.beam.sdk.values.{KV, PCollection, PCollectionView}
+import org.slf4j.LoggerFactory
 
 import scala.reflect.ClassTag
 
+private object PairSCollectionFunctions {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  case class BFSettings(width: Int, capacity: Int, numBFs: Int)
+
+  def optimalBFSettings(numEntries: Long, fpProb: Double): BFSettings = {
+    val maxTotalBytes = 1024L * 1024L * 1024L // 1GB
+
+    // double to int rounding error happens when numEntries > (1 << 27)
+    // set numEntries upper bound to 1 << 27 to avoid high false positive
+    def estimateWidth(numEntries: Int, fpProb: Double): Int =
+      math.ceil(-1 * numEntries * math.log(fpProb) / math.log(2) / math.log(2)).toInt
+
+    // upper bound of n as 2^x
+    def upper(n: Int): Int = 1 << (0 to 27).find(1 << _ >= n).get
+
+    // cap capacity between [minSize, maxSize] and find upper bound of 2^x
+    val (minSize, maxSize) = (2048, 1 << 27)
+    var capacity = upper(math.max(math.min(numEntries, maxSize).toInt, minSize))
+
+    // find a width with the given capacity
+    var width = estimateWidth(capacity, fpProb)
+    while (width == Int.MaxValue) {
+      capacity = capacity >> 1
+      width = estimateWidth(capacity, fpProb)
+    }
+    val numBFs = (numEntries / capacity).toInt + 1
+
+    logger.info(s"Optimal Bloom Filter settings for numEntries = $numEntries, fpProb = $fpProb")
+    logger.info(s"BF width = $width, capacity = $capacity, numBFs = $numBFs")
+    val totalBytes = width.toLong * numBFs / 8
+    if (totalBytes > maxTotalBytes) {
+      logger.warn(s"Total BF bytes $totalBytes > 1GB")
```

Comment:
Yeah, now i see how this is gonna be kinda arbitrary warning. Essentially it doesn't really matter from the perspective of this  algo, +/- how well partitioning works. Does this warning actually make sense at the end of the day? It doesn't suggest what to do, nor what is the problem.